src/Microsoft.AspNet.Server.Kestrel/Http/SocketOutput.cs

```diff
@@ -80,23 +78,25 @@ public SocketOutput(KestrelThread thread, UvStreamHandle socket, long connection
                 {
                     _nextWriteContext.SocketDisconnect = true;
                 }
+
                 // Complete the write task immediately if all previous write tasks have been completed,
                 // the buffers haven't grown too large, and the last write to the socket succeeded.
-                triggerCallbackNow = _lastWriteError == null &&
-                                     _callbacksPending.Count == 0 &&
-                                     _numBytesPreCompleted + buffer.Count <= _maxBytesPreCompleted;
-                if (triggerCallbackNow)
+                if (_lastWriteError == null &&
+                    _tasksPending.Count == 0 &&
+                    _numBytesPreCompleted + buffer.Count <= _maxBytesPreCompleted)
                 {
                     _numBytesPreCompleted += buffer.Count;
                 }
+                else if (immediate)
+                {
+                    // immediate write, which is not eligable for instant completion above
+                    tcs = new TaskCompletionSource<object>(buffer.Count);
+                    _tasksPending.Enqueue(tcs);
+                }
                 else
                 {
-                    _callbacksPending.Enqueue(new CallbackContext
-                    {
-                        Callback = callback,
-                        State = state,
-                        BytesToWrite = buffer.Count
-                    });
+                    // immediate==false calls always return complete tasks, because there is guaranteed
+                    // to be a subsequent immediate==true call which will go down one of the previous code-paths
```

Comment:
I think this is a logic bug. Since this write isn't being tracked in the `_tasksPending` queue, you need to consider this write as pre-completed (`_numBytesPreCompleted += buffer.Count`).

This bug could cause `_numBytesPreCompleted` to become negative leading to assertion failures in debug builds. In release builds, this could delay write throttling.

I think our existing SocketOutputTests don't catch this bug because they were written before the `immediate` parameter was introduced. As a result, the tests always use the default value for `immediate` which is true.

Once you fix this, could you add a regression test to SocketOutputTests?